compiler/rustc_data_structures/src/fingerprint.rs

```diff
@@ -7,19 +7,30 @@ use std::hash::{Hash, Hasher};
 use std::mem::{self, MaybeUninit};
 
 #[derive(Eq, PartialEq, Ord, PartialOrd, Debug, Clone, Copy)]
+#[repr(C)]
 pub struct Fingerprint(u64, u64);
 
 impl Fingerprint {
     pub const ZERO: Fingerprint = Fingerprint(0, 0);
 
+    #[inline]
+    pub fn new(_0: u64, _1: u64) -> Fingerprint {
+        Fingerprint(_0, _1)
+    }
+
     #[inline]
     pub fn from_smaller_hash(hash: u64) -> Fingerprint {
         Fingerprint(hash, hash)
     }
 
     #[inline]
     pub fn to_smaller_hash(&self) -> u64 {
-        self.0
+        // Even though both halves of the fingerprint are expected to be good
+        // quality hash values, let's still combine the two values because the
+        // Fingerprints in DefPathHash have the StableCrateId portion which is
+        // the same for all DefPathHashes from the same crate. Combining the
+        // two halfs makes sure we get a good quality hash in such cases too.
+        self.0.wrapping_mul(3).wrapping_add(self.1)
```

Comment:
Hmm. The previous version of the code had an invariant where for all u64 `h`, `Fingerprint::from_smaller_hash(h).to_smaller_hash() == h`, and this new version does not support that invariant (if I understand correctly).

But I suppose that is also ... the point? (One workaround would be to make `from_smaller_hash` set `self.0` to 0. But I would expect that to have bad fallout elsewhere...)

I don't know if that invariant *matters*, but I'd be careful about breaking it, since its the kind of thing I would *expect* to hold...

Reply:
Okay I reviewed the code, and `from_smaller_hash` is only called in one place, on `local_id.as_u32()`. I don't think we're making use of the invariant that I wrote above.